lib/active_storage/service/cloudinary_service.rb

```diff
@@ -0,0 +1,190 @@
+require 'active_storage/blob_key'
+require 'cloudinary/helper'
+
+unless ActiveStorage::Blob.method_defined? :original_key
+  class ActiveStorage::Blob
+    alias_method :original_key, :key
+
+    def key
+      original_key
+      ActiveStorage::BlobKey.new(@attributes.as_json)
+    end
+  end
+end
+
+module ActiveStorage
+  class Service::CloudinaryService < Service
+    module Headers
+      CONTENT_TYPE = "Content-Type".freeze
+      CONTENT_MD5 = "Content-MD5".freeze
+    end
+    attr_reader :upload_options
+
+    def initialize(**options)
+      @options = options
+      @helper = ActionView::Base.new
+    end
+
+    def upload(key, io, filename: nil, checksum: nil, **options)
+      instrument :upload, key: key, checksum: checksum do
+        begin
+          extra_headers = checksum.nil? ? {} : {Headers::CONTENT_MD5 => checksum}
+          options = @options.merge(options)
+          Cloudinary::Uploader.upload(
+            io,
+            public_id: public_id(key),
+            resource_type: resource_type(io, key),
+            context: {active_storage_key: key, checksum: checksum},
+            extra_headers: extra_headers,
+            **options
+          )
+        rescue CloudinaryException
+          raise ActiveStorage::IntegrityError
```

Comment:
Might make sense to add a message or wrap the `CloudinaryException` or at least pass on the backtrace?